RecoTracker/PixelSeeding/plugins/alpaka/CAHitNtuplet.cc

```diff
@@ -20,57 +22,230 @@
+      // of modules each layer has. And we need the  extra spot
+      // at the end to hold the total number of modules.
+
+      for (auto& det : dets) {
+        DetId detid = det->geographicalId();
+#ifdef GPU_DEBUG
+        if (n_modules >= int(subSystemOffset)) {
+          subSystemName = GeomDetEnumerators::tkDetEnum[++subSystem];
+          subSystemOffset = trackerGeometry.offsetDU(subSystemName);
+          std::cout << " ===================== Subsystem: " << subSystemName << std::endl;
+        }
+#endif
+
+        auto layer = trackerTopology.layer(detid);
+
+        if (layer != oldLayer) {
+          layerStarts[layerCount++] = n_modules;
+
+          if (layerCount > n_layers + 1)
+            break;
+
+          oldLayer = layer;
+#ifdef GPU_DEBUG
+          std::cout << " > New layer at module : " << n_modules << " (detId: " << detid << ")" << std::endl;
+#endif
+        }
+
+        n_modules++;
+      }
+
+      reco::CAGeometryHost product{{{n_layers + 1, n_pairs, n_modules}}, cms::alpakatools::host()};
+
+      auto layerSoA = product.view();
+      auto cellSoA = product.view<::reco::CAGraphSoA>();
+      auto modulesSoA = product.view<::reco::CAModulesSoA>();
+
+      for (int i = 0; i < n_modules; ++i) {
+        auto det = dets[i];
+        auto vv = det->surface().position();
+        auto rr = Rotation(det->surface().rotation());
+        modulesSoA[i].detFrame() = Frame(vv.x(), vv.y(), vv.z(), rr);
+      }
+
+      for (int i = 0; i < n_layers; ++i) {
+        layerSoA.layerStarts()[i] = layerStarts[i];
+        layerSoA.caThetaCut()[i] = iCache->caThetaCuts_[i];
+        layerSoA.caDCACut()[i] = iCache->caDCACuts_[i];
+      }
+
+      layerSoA.layerStarts()[n_layers] = layerStarts[n_layers];
+
+      for (int i = 0; i < n_pairs; ++i) {
+        cellSoA.graph()[i] = {{uint32_t(iCache->pairGraph_[2 * i]), uint32_t(iCache->pairGraph_[2 * i + 1])}};
+        cellSoA.phiCuts()[i] = iCache->phiCuts_[i];
+        cellSoA.minz()[i] = iCache->minZ_[i];
+        cellSoA.maxz()[i] = iCache->maxZ_[i];
+        cellSoA.maxr()[i] = iCache->maxR_[i];
+        cellSoA.startingPair()[i] = false;
+      }
+
+      for (const int& i : iCache->startingPairs_)
+        cellSoA.startingPair()[i] = true;
+
+      return std::make_shared<CAGeometryCache>(std::move(product));
+    }
+
+    static std::unique_ptr<::reco::CAGeometryParams> initializeGlobalCache(edm::ParameterSet const& iConfig) {
+      return std::make_unique<::reco::CAGeometryParams>(iConfig.getParameterSet("geometry"));
+    }
+
   private:
     const edm::ESGetToken<MagneticField, IdealMagneticFieldRecord> tokenField_;
-    const device::ESGetToken<PixelCPEFastParams<TrackerTraits>, PixelCPEFastParamsRecord> cpeToken_;
     const device::EDGetToken<HitsOnDevice> tokenHit_;
     const device::EDPutToken<TkSoADevice> tokenTrack_;
 
+    const TFormula maxNumberOfDoublets_;
+    const TFormula maxNumberOfTuples_;
```

Comment:
In the (distant?) past `TFormula` has had threading problems (at least performance, not sure about safety). I'd suggest to look into `reco::FormulaEvaluator` instead
https://github.com/cms-sw/cmssw/blob/ea8a8a7984665c7013c922b8a0ee0b30eb1b1ee9/CommonTools/Utils/interface/FormulaEvaluator.h#L69

See also https://github.com/cms-sw/cmssw/pull/11516

Reply:
Thanks I'm looking into it.

Reply:
Updated to use the `FormulaEvaluator`.